framework/operator_test.go

```diff
@@ -0,0 +1,364 @@
+	m := "ProcessDeleteState"
+	r.Order = append(r.Order, m)
+
+	if r.returnErrorFor(m) {
+		return r.Error
+	}
+
+	return nil
+}
+
+func (r *testResource) returnErrorFor(errorMethod string) bool {
+	ok := r.Error != nil && r.ErrorCount > r.errorCount && r.ErrorMethod == errorMethod
+
+	if ok {
+		r.errorCount++
+		return true
+	}
+
+	return false
+}
+
+// Test_Operator_ProcessCreate_NoResource ensures there is an error thrown when
+// executing ProcessCreate without having any resources provided.
+func Test_Operator_ProcessCreate_NoResource(t *testing.T) {
+	err := testMustNewFramework(t).ProcessCreate(nil, nil)
+	if !IsExecutionFailed(err) {
+		t.Fatal("expected", true, "got", false)
+	}
+}
+
+// Test_Operator_ProcessDelete_NoResource ensures there is an error thrown when
+// executing ProcessDelete without having any resources provided.
+func Test_Operator_ProcessDelete_NoResource(t *testing.T) {
+	err := testMustNewFramework(t).ProcessDelete(nil, nil)
+	if !IsExecutionFailed(err) {
+		t.Fatal("expected", true, "got", false)
+	}
+}
+
+// Test_Operator_ProcessCreate_ResourceOrder ensures the resource's methods are
+// executed as expected when creating resources.
+func Test_Operator_ProcessCreate_ResourceOrder(t *testing.T) {
+	tr := &testResource{}
+	rs := []Resource{
+		tr,
+	}
+
+	err := testMustNewFramework(t).ProcessCreate(nil, rs)
+	if err != nil {
+		t.Fatal("expected", nil, "got", err)
+	}
+
+	e := []string{
+		"GetCurrentState",
+		"GetDesiredState",
+		"GetCreateState",
+		"ProcessCreateState",
+	}
+	if !reflect.DeepEqual(e, tr.Order) {
+		t.Fatal("expected", e, "got", tr.Order)
+	}
+}
+
+// Test_Operator_ProcessCreate_ResourceOrder_Retry ensures the resource's
+// methods are executed as expected when retrying the creation process.
+func Test_Operator_ProcessCreate_ResourceOrder_Retry(t *testing.T) {
+	testCases := []struct {
+		ErrorCount          int
+		ErrorMethod         string
+		ExpectedMethodOrder []string
+	}{
+		{
+			ErrorCount:  1,
+			ErrorMethod: "GetCurrentState",
+			ExpectedMethodOrder: []string{
+				"GetCurrentState",
+				"GetCurrentState",
+				"GetDesiredState",
+				"GetCreateState",
+				"ProcessCreateState",
+			},
+		},
+		{
+			ErrorCount:  2,
+			ErrorMethod: "GetCurrentState",
+			ExpectedMethodOrder: []string{
+				"GetCurrentState",
+				"GetCurrentState",
+				"GetCurrentState",
+				"GetDesiredState",
+				"GetCreateState",
+				"ProcessCreateState",
+			},
+		},
+		{
+			ErrorCount:  2,
+			ErrorMethod: "ProcessCreateState",
+			ExpectedMethodOrder: []string{
+				"GetCurrentState",
+				"GetDesiredState",
+				"GetCreateState",
+				"ProcessCreateState",
+				"ProcessCreateState",
+				"ProcessCreateState",
+			},
+		},
+	}
+
+	for i, tc := range testCases {
+		tr := &testResource{
+			Error:       executionFailedError,
+			ErrorCount:  tc.ErrorCount,
+			ErrorMethod: tc.ErrorMethod,
+		}
+		rs := []Resource{
+			tr,
+		}
+		bf := func() backoff.BackOff {
+			return &backoff.ZeroBackOff{}
```

Comment:
The zero backoff retries immediately. This is cool for testing. I initially used the exponential backoff. The tests took around 5 seconds. Now they take around 7 milliseconds. I felt super smart when I realized this. :D